aws/hll/aws-s3-transfer-manager/src/download/body.rs

```diff
@@ -0,0 +1,223 @@
+        // TODO(aws-sdk-rust#1159, design) - do we want ChunkResponse (or similar) rather than AggregatedBytes? Would
+        //  make additional retries of an individual chunk/part more feasible (though theoretically already exhausted retries)
+        loop {
+            if self.sequencer.is_ordered() {
+                break;
+            }
+
+            let chunk = self.inner.next().await;
+            if chunk.is_none() {
+                break;
+            }
+
+            match chunk? {
+                Ok(chunk) => self.sequencer.push(chunk),
+                Err(err) => return Some(Err(err)),
+            }
+        }
+
+        let chunk = self
+            .sequencer
+            .pop()
+            .map(|r| Ok(r.data.expect("chunk data")));
+
+        if chunk.is_some() {
+            // if we actually pulled data out, advance the next sequence we expect
+            self.sequencer.advance();
+        }
+
+        chunk
+    }
+}
+
+#[derive(Debug)]
+struct Sequencer {
+    /// next expected sequence
+    next_seq: u64,
+    chunks: BinaryHeap<cmp::Reverse<SequencedChunk>>,
+}
+
+impl Sequencer {
+    fn new() -> Self {
+        Self {
+            chunks: BinaryHeap::with_capacity(8),
+            next_seq: 0,
+        }
+    }
+
+    fn push(&mut self, chunk: ChunkResponse) {
+        self.chunks.push(cmp::Reverse(SequencedChunk(chunk)))
+    }
+
+    fn pop(&mut self) -> Option<ChunkResponse> {
+        self.chunks.pop().map(|c| c.0 .0)
+    }
+
+    fn is_ordered(&self) -> bool {
+        let next = self.peek();
+        if next.is_none() {
+            return false;
+        }
+
+        next.unwrap().seq == self.next_seq
+    }
+
+    fn peek(&self) -> Option<&ChunkResponse> {
+        self.chunks.peek().map(|c| &c.0 .0)
+    }
+
+    fn advance(&mut self) {
+        self.next_seq += 1
+    }
+}
+
+#[derive(Debug)]
+struct SequencedChunk(ChunkResponse);
+
+impl Ord for SequencedChunk {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.seq.cmp(&other.0.seq)
+    }
+}
+
+impl PartialOrd for SequencedChunk {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for SequencedChunk {}
+impl PartialEq for SequencedChunk {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.seq == other.0.seq
+    }
+}
+
+/// A body that returns chunks in whatever order they are received.
+#[derive(Debug)]
+pub(crate) struct UnorderedBody {
+    chunks: Option<mpsc::Receiver<Result<ChunkResponse, TransferError>>>,
+}
+
+impl UnorderedBody {
+    fn new(chunks: Option<BodyChannel>) -> Self {
+        Self { chunks }
+    }
+
+    /// Pull the next chunk of data off the stream.
+    ///
+    /// Returns [None] when there is no more data.
+    /// Chunks returned from an [UnorderedBody] are not guaranteed to be sequenced
+    /// in the right order. Consumers are expected to sequence the data themselves
```

Comment:
I think we should prefer the term `sorted` to `sequenced` b/c `sorted` feels more common to me.